[https://nvbugs/5593199][test] Enhance beam search tests deterministic dummy model

stnie opened the pull request:
- added copyright header
- Introduced a DummyModel and associated configurations to facilitate deterministic outputs for beam search tests.
- Updated test cases to validate generation logits, logprobs, and cache indirection.
- Refactored input prompts and expected outputs to align with the new model structure.
- Added checks for context logits and overall output validation to ensure correctness in beam search functionality.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Tests**
  * Enhanced beam search test infrastructure with improved scaffolding and deterministic utilities for more robust and comprehensive test coverage.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

<!--
Please write the PR title by following this template:

**[JIRA ticket/NVBugs ID/GitHub issue/None][type] Summary**

Valid ticket formats:
  - JIRA ticket: [TRTLLM-1234] or [FOOBAR-123] for other FOOBAR project
  - NVBugs ID: [https://nvbugs/1234567]
  - GitHub issue: [#1234]
  - No ticket: [None]

Valid types (lowercase): [fix], [feat], [doc], [infra], [chore], etc.

Examples:
  - [TRTLLM-1234][feat] Add new feature
  - [https://nvbugs/1234567][fix] Fix some bugs
  - [#1234][doc] Update documentation
  - [None][chore] Minor clean-up

Alternative (faster) way using CodeRabbit AI:

**[JIRA ticket/NVBugs ID/GitHub issue/None] @coderabbitai title**

NOTE: "@coderabbitai title" will be replaced by the title generated by CodeRabbit AI, that includes the "[type]" and title.
For more info, see /.coderabbit.yaml.

-->

## Description

Updated beam search tests to provide deterministic output, to prevent the choice of hardware to affect the results

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

stnie committed:
[https://nvbugs/5593199][test] Enhance beam search tests with dummy model and validation checks to provide deterministic output

- added copyright header
- Introduced a DummyModel and associated configurations to facilitate deterministic outputs for beam search tests.
- Updated test cases to validate generation logits, logprobs, and cache indirection.
- Refactored input prompts and expected outputs to align with the new model structure.
- Added checks for context logits and overall output validation to ensure correctness in beam search functionality.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

Funatiq commented on tests/unittest/_torch/sampler/test_beam_search.py:
```diff
@@ -1,43 +1,278 @@
+            num_context_requests:num_requests].view(
+                num_generation_requests * attn_metadata.beam_width,
+                attn_metadata.cache_indirection.shape[-1]
+            )[:num_generation_beams]
```
`num_generation_beams = num_generation_requests * attn_metadata.beam_width`, so why do we need `[:num_generation_beams]` again?

stnie replied:
We don't. I'll remove this.

Funatiq commented on tests/unittest/_torch/sampler/test_beam_search.py:
```diff
@@ -1,43 +1,278 @@
+    top_offset, second_best_offset = torch.topk(softmax_score, k=2,
+                                                dim=-1).indices
+
+    # iteration 1
```
I suggest changing that since the loop below starts at 1.
```suggestion
    # First iteration
```

stnie commented on tests/unittest/_torch/sampler/test_beam_search.py:
```diff
@@ -1,43 +1,278 @@
+                expected_outputs[1][j] = expected_outputs[0][j]
+                if use_overlap_scheduler or i < num_iterations - 1:
+                    # Avoid swapping in the last iteration, as the cache indirection provided by the model does not reflect this change
+                    # in case of overlap scheduler one additional model forward is performed, so we do not need to skip the last swap
```
The `sampler` does perform the swap. But the model output is created before the sampling step in which this swap happens. So, the `cache indirection` from the model output does not "contain" any changes to `cache indirection`, which may happen during the following sampling step. Therefore, this code should also skip the last update to the `cache indirection`.

stnie replied:
I updated the comment and I also skipped adding anything to the cache indirection in the last step to make it clearer that the final cache indirection is not checked

stnie committed:
[https://nvbugs/5593199][chore] Refactor beam search test functions for clarity and correctness

- Removed reliance on additional output from overlap scheduler
- Removed unused parameters from `get_expected_outputs` and `validate_output` functions to simplify the interface.
- Updated comments for better clarity on the logic flow within the beam search output validation.
- Adjusted cache indirection checks to ensure accuracy in the last token generation.
- Enhanced readability by restructuring code and comments in the test cases.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #22650](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/22650/) [ run ] completed with state `SUCCESS`. Commit: `44df581` 
[/LLM/main/L0_MergeRequest_PR pipeline #17074](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/17074/) completed with status: 'FAILURE'
